File: backend/src/main/java/com/example/backend/controller/GroupController.java
package com.example.backend.controller;

import com.example.backend.model.Group;
import com.example.backend.service.GroupService;
import com.example.backend.service.PostService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

@RestController
@RequestMapping("/api/groups")
@CrossOrigin(origins = { "http://localhost:3000", "http://localhost:3001", "http://localhost:3002" })
public class GroupController {
    private static final Logger logger = LoggerFactory.getLogger(GroupController.class);

    @Autowired
    private GroupService groupService;

    @Autowired
    private PostService postService; // Add PostService dependency

    @PostMapping
    public ResponseEntity<?> createGroup(
            @RequestParam("name") String name,
            @RequestParam("description") String description,
            @RequestParam("userId") String userId,
            @RequestParam(value = "coverImage", required = false) MultipartFile coverImage) {
        try {
            logger.info("Creating group: {}", name);
            if (name == null || name.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("Group name is required");
            }
            if (userId == null || userId.trim().isEmpty()) {
                return ResponseEntity.badRequest().body("User ID is required");
            }

            Group group = groupService.createGroup(name, description, userId, coverImage);
            return ResponseEntity.ok(group);
        } catch (IllegalArgumentException e) {
            logger.error("Invalid group data: {}", e.getMessage());
            return ResponseEntity.badRequest().body(e.getMessage());
        } catch (Exception e) {
            logger.error("Error creating group: {}", e.getMessage());
            return ResponseEntity.badRequest()
                    .body("Failed to create group: " + e.getMessage());
        }
    }

    @GetMapping
    public ResponseEntity<?> getAllGroups() {
        try {
            return ResponseEntity.ok(groupService.getAllGroups());
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/user/{userId}")
    public ResponseEntity<?> getUserGroups(@PathVariable String userId) {
        try {
            return ResponseEntity.ok(groupService.getUserGroups(userId));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/{groupId}/posts")
    public ResponseEntity<?> getGroupPosts(@PathVariable String groupId) {
        try {
            return ResponseEntity.ok(postService.getGroupPosts(groupId));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        }
    }

    @GetMapping("/{groupId}")
    public ResponseEntity<?> getGroupById(@PathVariable String groupId) {
        try {
            logger.info("Fetching group: {}", groupId);
            Group group = groupService.getGroupById(groupId);
            return ResponseEntity.ok(group);
        } catch (IllegalArgumentException e) {
            logger.error("Group not found: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
        } catch (Exception e) {
            logger.error("Error fetching group: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to fetch group");
        }
    }

    @PutMapping("/{groupId}")
    public ResponseEntity<?> updateGroup(
            @PathVariable String groupId,
            @RequestParam String userId,
            @RequestParam String name,
            @RequestParam String description) {
        try {
            Group updatedGroup = groupService.updateGroup(groupId, userId, name, description);
            return ResponseEntity.ok(updatedGroup);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to update group: " + e.getMessage());
        }
    }

    @DeleteMapping("/{groupId}")
    public ResponseEntity<?> deleteGroup(
            @PathVariable String groupId,
            @RequestParam String userId) {
        try {
            groupService.deleteGroup(groupId, userId);
            return ResponseEntity.ok().build();
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(e.getMessage());
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to delete group: " + e.getMessage());
        }
    }
}

File: backend/src/main/java/com/example/backend/controller/groupController.js
const Group = require('../models/Group');
const User = require('../models/User');

// Create a new group
exports.createGroup = async (req, res) => {
  try {
    const { name, description, members, admin } = req.body;
    
    if (!name || !admin) {
      return res.status(400).json({ message: 'Group name and admin are required' });
    }

    const newGroup = new Group({
      name,
      description,
      members: members || [],
      admin
    });

    // Add admin to members if not already included
    if (!newGroup.members.includes(admin)) {
      newGroup.members.push(admin);
    }

    const savedGroup = await newGroup.save();
    res.status(201).json(savedGroup);
  } catch (error) {
    console.error('Error creating group:', error);
    res.status(500).json({ message: 'Server error while creating group', error: error.message });
  }
};

// Get all groups
exports.getAllGroups = async (req, res) => {
  try {
    const groups = await Group.find();
    res.status(200).json(groups);
  } catch (error) {
    console.error('Error fetching groups:', error);
    res.status(500).json({ message: 'Server error while fetching groups' });
  }
};

// Get group by ID
exports.getGroupById = async (req, res) => {
  try {
    const group = await Group.findById(req.params.id);
    if (!group) {
      return res.status(404).json({ message: 'Group not found' });
    }
    res.status(200).json(group);
  } catch (error) {
    console.error('Error fetching group:', error);
    res.status(500).json({ message: 'Server error while fetching group' });
  }
};
